File: spatialsegregation/src/isar.cpp
#include "isar.h"
/**********************************************************************************/
std::vector<double> isar(Graph *graph, double *fpar, int *dbg, int *included)
{
	if(fpar[1]==1) return isar_wdeg(graph, fpar, dbg, included);
	else if(fpar[1]==2) return isar_markweighted(graph, fpar, dbg, included);
	else if(fpar[1]==3) return isar_normal_empty(graph, fpar, dbg, included);
	return isar_normal(graph, fpar, dbg, included);
}
/*********************************************************************************/
std::vector<double> isar_normal(Graph *graph, double *fpar, int *dbg, int *included)
{
	if(*dbg)Rprintf("isar[type=%i",(int) fpar[0]);
	int i, j, k,l, dbg0,nt=0;
	double a1[1], m;
	std::vector<double> value;
	value.clear();

	if((int)fpar[0]<=0)
	{
		dbg0 = *dbg;
		*dbg = 0;
		for(i=0;i< graph->pp->getNtypes();i++)
		{
			if(graph->typeIncluded.at(i))
			{
				nt++;
				a1[0] = (double) graph->pp->getTypevec(&i);
				value.push_back(isar_normal(graph,a1,dbg,included).at(0));
			}
		}
//		Rprintf("nt=%i,",nt);
		*dbg = dbg0;
	}
	else
	{
		int target_type = (int) fpar[0];
		value.push_back(0.0);
		int n=0,typej;

		for(i=0;i< (int)graph->nodelist.size();i++) //sum over target species...
		{
			if(included[i] && graph->pp->getT(&i)==target_type) //...
			{
				n++; // one more of target_type
				m=0; // how many types around i?
				for(j=0; j< graph->pp->getNtypes();j++) // sum over all species
				{
					typej = graph->pp->getTypevec(&j);
					for(k=0;k < (int)graph->nodelist[i].size();k++)  // check if type j present
					{
						l = graph->nodelist[i][k]-1;
						if(graph->pp->getT(&l)==typej) // if i's neighbour k is of type j
						{
							m = m + graph->getTypeToTypeWeight(&target_type, &typej);
							break;
						}
					}
				}
				value.at(0)=value.at(0)+ m;
				graph->pp->setMass2(&i,&m);
			}
		}
		if(n>0) value.at(0) = value.at(0)/(double)n; // mean value over the species
	}

	if(*dbg)Rprintf("]");
	return value;
}

/*********************************************************************************/
std::vector<double> isar_normal_empty(Graph *graph, double *fpar, int *dbg, int *included)
{
	if(*dbg)Rprintf("isar(empty)[type=%i",(int) fpar[0]);
	int i, j, k,l, dbg0;
	double a1[1], m;
	std::vector<double> value;
	value.clear();

	if((int)fpar[0]<=0)
	{
		dbg0 = *dbg;
		*dbg = 0;
		for(i=0;i< graph->pp->getNtypes();i++)
		{
			if(graph->typeIncluded.at(i))
			{
				a1[0] = (double) graph->pp->getTypevec(&i);
				value.push_back(isar_normal_empty(graph,a1,dbg,included).at(0));
			}
		}
		*dbg = dbg0;
	}
	else
	{
		int target_type = (int) fpar[0];
		value.push_back(0.0);
		int n, typej, empty;
		for(j=0; j< graph->pp->getNtypes();j++) // sum P_tj over j
		{
			n=0;
			m=0.0;
			typej = graph->pp->getTypevec(&j);
			for(i=0;i< (int)graph->nodelist.size();i++) //sum over target species to estimate P_tj...
			{
				if(included[i] && graph->pp->getT(&i)==target_type) //...
				{
					n++; // one more of target_type
					empty = 1;
					for(k=0;k < (int)graph->nodelist[i].size();k++)  // check if type j present
					{
						l = graph->nodelist[i][k]-1;
						if(graph->pp->getT(&l)==typej) // if i's neighbour k is of type j
						{
							empty = 0;
							break;
						}
					}
					m = m + empty;
				}
			}
			if(n>0) m = m/(double)n;
			value.at(0)=value.at(0)+ (1.0-m);
			graph->pp->setMass2(&i,&m);
		}
	}
	if(*dbg)Rprintf("]");
	return value;
}

/**********************************************************************************/
std::vector<double> isar_wdeg(Graph *graph, double *fpar, int *dbg, int *included)
{
	if(*dbg)Rprintf("isar (degree weighted)[type=%i",(int) fpar[0]);
	int i, j, n, k,l, dbg0;
	double m;
	double a1[1],vi;
	std::vector<double> value;
	value.clear();

	if((int)fpar[0]<=0)
	{
		dbg0 = *dbg;
		*dbg = 0;
		for(i=0;i< graph->pp->getNtypes();i++)
		{
			if(graph->pp->lambdas[i]>0)
			{
				a1[0] = (double) graph->pp->getTypevec(&i);
				value.push_back(isar_wdeg(graph,a1,dbg,included).at(0));
			}
		}
		*dbg = dbg0;
	}
	else
	{
		int target_type = (int) fpar[0];
		int typej=0;
		value.push_back(0.0);
		n=0;
		for(i=0;i< (int)graph->nodelist.size();i++) //sum over target species...
		{
			if(included[i] && graph->pp->getT(&i)==target_type) //...
			{
				m=0; // how many types around i?
				n++;
				for(j=0; j< graph->pp->getNtypes();j++) // sum over all species
				{
					typej = graph->pp->getTypevec(&j);
					for(k=0;k < (int)graph->nodelist[i].size();k++)  // check if type j present
					{
						l = graph->nodelist[i][k]-1;
						if(graph->pp->getT(&l)==typej) // if i's neighbour k is of type j
						{
							m=m+graph->getTypeToTypeWeight(&target_type, &typej);
							break;
						}
					}
				}
				if(graph->nodelist[i].size()>0){
					vi = (double)m/graph->nodelist[i].size();
					value.at(0)=value.at(0)+ vi;
					graph->pp->setMass2(&i, &vi);
				}else n--;
			}
		}
		if(n>0) value.at(0) = value.at(0)/(double)n;
	}

	if(*dbg)Rprintf("]");
	return value;
}


/**********************************************************************************/
// not complete.
std::vector<double> isar_markweighted(Graph *graph, double *fpar, int *dbg, int *included)
{
	if(*dbg)Rprintf("isar (mass weighted)[type=%i",(int) fpar[0]);
	int i, j, n, k, l, dbg0;
	double a1[1], m;
	std::vector<double> value;
	value.clear();

	if((int)fpar[0]<=0)
	{
		dbg0 = *dbg;
		*dbg = 0;
		for(i=0;i< graph->pp->getNtypes();i++)
		{
			if(graph->pp->lambdas[i]>0)
			{
				a1[0] = (double) graph->pp->getTypevec(&i);
				value.push_back(isar_markweighted(graph,a1,dbg,included).at(0));
			}
		}
		*dbg = dbg0;
	}
	else
	{
		int target_type = (int) fpar[0];
		int isnew, typej;
		value.push_back(0.0);
		n=0;
		std::vector<int> p;
		std::vector<int> counts;
		std::vector<double> M;

		for(i=0;i< (int)graph->nodelist.size();i++) //sum over target species...
		{
			m=0;
			if(included[i] && graph->pp->getT(&i)==target_type) //...
			{
				p.clear();
				counts.clear();
				M.clear();
				n++;
				for(j=0; j < (int)graph->nodelist.at(i).size();j++)  // check if type j present
				{
					isnew=1;
					l = graph->nodelist[i][j]-1;
					typej = graph->pp->getT(&l);
					for(k=0; k < (int) p.size(); k++)
					{
						if(typej == p.at(k))
						{
							counts.at(k)++;
							M.at(k) = M.at(k) + graph->pp->getMass(&l)*graph->getTypeToTypeWeight(&target_type, &typej);
							isnew = 0;
							break;
						}
					}
					if(isnew)
					{
						counts.push_back(1);
						l = graph->nodelist.at(i).at(j)-1;
						M.push_back(graph->pp->getMass(&l));
						p.push_back(typej);
					}
				}
				for(k=0; k < (int)p.size();k++)
					m = m + (double) M.at(k)/(double)counts.at(k);
				value.at(0)=value.at(0)+(double)m;
			}
			graph->pp->setMass2(&i, &m);

		}
		value.at(0) = value.at(0)/(double)n;

	}

	if(*dbg)Rprintf("]");
	return value;
}

File: spatialsegregation/R/isar.R
# Last update: 090715
###############################################################################

#' Individual Species Area Relationship
#' 
#' Compute the Individual Species Area Relationship ( ISAR ) or Local Species
#' Richness, for a given multitype point pattern.
#' @param X Multitype point pattern of class \code{ppp} (see package 'spatstat')
#' @param r Vector of sizes for neighbourhoods, e.g. \code{geometric} graph with
#'   different ranges.
#' @param target Default NULL. Calculate only for target type. If NULL computes
#'   for each type + mean over all types.
#' @param v2 Logical. Estimate species-to-neighbours-ratio instead of just total
#'   number of species.
#' @param v3 Logical. Instead of summing number 1 for each species present, sum
#'   the average X$mass of each species present.
#' @param v4 Logical. Estimate ISAR using empty space probabilities instead of
#'   direct counts (equals the normal version in all my tests)
#' @param ntype Sets the n'hood type to \code{knn} by default in isar.index.
#' @param ... Further parameters for the function \code{segregationFun}.
#'   
#' @details Extension of ISAR-function introduced in WGGH07. In effect
#' calculates the expected amount of different types present in the
#' neighbourhood of a point in the pattern.
#' 
#' The function \code{isarF} is the calculation function for different
#' neighbourhoods. Uses function \code{\link{segregationFun}}.
#' 
#' The function \code{isar.index} is a shortcut to get a single value for the
#' pattern. Uses 4-nn graph by default.
#' 
#' @references 
#' Rajala, Illian: A family of spatial biodiversity measures based on graphs, Env. Ecol. Stat. 2012
#' 
#' Wiegand, Gunatilleke, Gunatilleke, Huth: How individual species structure diversity in tropical forests. PNAS, nov 16, 2007. 
#' 
#' @export
isarF<-function(X, r=NULL, target=NULL, v2=FALSE, v3=FALSE, v4=FALSE, ... )
{
	# check that X is ppp-object
	verifyclass(X, "ppp")
	if(length(levels(X$marks))<2) stop("Use only on a multitype point pattern (data.frame-marks not yet supported).")
	# if no target given, calculate for all types
	if(is.null(target))
	{
		targeti <- 0
		valu   <- "ISARmean"
	}
	# else convert to an integer
	else
	{
		if(!is.factor(X$marks))warning("Marks of X are not in factor form. Transforming.")
		X$marks<-as.factor(X$marks)
		targeti<- which( levels(X$marks)  == target)
#		targeti<-which( union(X$marks, NULL) == target)
		if(length(targeti)!=1) stop("Target type not one of pattern types.")
	}
	
	#v2 logical if a degree weighted version should be calculated
	if(v2) funtype <- "Neighbour-count-weighted-ISAR"
	if(v3){v2<-2; funtype <- "Mass-weighted ISAR"; }
	if(v4){v2<-3; funtype <- "eISAR"; }
	else funtype <- "ISAR"
	
		
	# use the main calc function
	res<-segregationFun(X=X, fun="isar", r, funpars=c(targeti, as.integer(v2)), ...)
	
	# theoretical values in CSR: depends on the neighbourhood type
	ntype<-res$ntype
	mdeg<-function(l,k)c( pi*l*k^2, k, 4, 6)[charmatch(ntype, kGraphs)]
	
	# get the intensities
	sum0<-summary(X)
	omarks<-order(union(X$marks,NULL)) # right order of marks
	l<-sum0$marks[,3][omarks]
	#    calc the theoretical values, also for the degree weighted version
	theo<-NULL
	for(para in res$parvec)theo<-
				c(theo, 
                 sum(1-exp(-mdeg(sum(l),para)*l/sum(l) )) / ifelse(v2,mdeg(sum(l), para),1))
										
		
	# create the fv-object
	isar.final<-fv(data.frame(theo=theo,par=res$parvec), 
			       argu="par",
				   alim=range(res$parvec),
				   ylab=substitute(ISAR, NULL),
			       desc=c("CSR values","Parameter values"),
				   valu="theo",
				   fmla=".~par",
			   unitname=res$unitname,
	   			  fname=funtype
			           )
					   
	# add all typewise values if no target type given
	if(targeti==0)
	{
		# the values from calculation
		tw<-res$v
		# set the names right, and don't forget to check inclusion (might drop some types off)
		colnames(tw)<-union(marks(X[res$included]),NULL)
		isar.final<-bind.fv(x=isar.final,
						    y=tw,
					  	 desc=paste("Typewise",funtype,"for type",colnames(tw)),
					     labl=colnames(tw)
					         )
					
		isar.final<-bind.fv(x=isar.final,
				            y=data.frame("ISARmean"=apply(res$v,1,mean,na.rm=TRUE)),
				         desc=paste("Mean",funtype,"over types"),
						 labl="ISARmean",
						 preferred="ISARmean"
				 	     	 )		
		# a frequency weighted mean instead of just a mean, w=freqs/sum(freqs)
		#Iw=apply(res$v,2,weighted.mean,w=w,na.rm=TRUE), 
	}
	
	# if target type given add the values for the target type
	else
	{
		isar.final<-bind.fv(x=isar.final,
				            y=data.frame("ISAR"=res$v[,1]),
						 desc=paste(funtype,"for type",target),
						 labl="ISAR",
					preferred="ISAR"
								)
	}

	# attach the frequencies too
	attr(isar.final,"frequencies")<-freqs(X[res$included])
	
	# and some notes
	attr(isar.final,"neighbourhoodType")<-res$ntype
	attr(isar.final,"note")<-res$note
	# add pointwise values
	attr(isar.final,"point.values")<-res$point.values
	# return 
	isar.final
}

###############################################################################
#' @export
#' @describeIn isarF Shortcut for 4-nearest neighbour value.
isar.index<-function(X, r=4, ntype="knn", ...)
{
	if(length(r)>1)stop("Use isarF for vector of parameter values.")
	I0<-isarF(X, r=r, ntype=ntype, ...)
	data.frame(meanISAR=I0$I, par=I0$par)
}